tests/ga/test_update.py

```diff
@@ -2110,6 +2110,141 @@ class ReportStatusTestCase(AgentTestCase):
+
+        for p in self.patches:
+            p.start()
+
+        return AgentTestCase.setUp(self)
+
+    def tearDown(self):
+        
+        for p in self.patches:
+            p.stop()
+
+        return AgentTestCase.tearDown(self)
+
+    @staticmethod
+    @contextlib.contextmanager
+    def _mock_update_handler(iterations=1, **kwargs):
+        """
+        Creates an UpdateHandler instance that will run n iterations.
+
+        Can be supplied keyword args for:
+            *   mock_wire_data_file: This arg is treated like mock_wire_protocol
+                would. Defaults to mockwiredata.DATA_FILE_STATUS_BLOB
+            *   http_<action>_handler where action is get, put, or post: This arg
+                is treated like mock_wire_protocol would.
+        
+        Returned UpdateHandler instance has its protocol mocked via mock_wire_protocol
+        and its _report_status function wrapped in a mock for accounting purposes.
+        """
+        
+        # Build the side_effect list for the UpdateHandler.is_running PropertyMock.
+        # Return True for the first n iterations followed by a single False to stop
+        # and then another True because the current UpdateHanlder implementation 
+        # does a __set__ during shutdown.
+        is_running_return_values = [True] * iterations + [False, True]
+        is_running_patch = patch.object(UpdateHandler, "is_running", PropertyMock(side_effect=is_running_return_values))
+
+        mock_http_get = kwargs.get("http_get_handler")
+        mock_http_put = kwargs.get("http_put_handler")
+        mock_http_post = kwargs.get("http_post_handler")
+
+        mock_wire_data_file = kwargs.get("mock_wire_data_file", mockwiredata.DATA_FILE_STATUS_BLOB)
+
+        try:
+            with mock_wire_protocol(mock_wire_data_file, mock_http_get, mock_http_post, mock_http_put) as protocol:
+                update_handler = get_update_handler()
+                
+                update_handler.protocol_util.get_protocol = Mock(return_value=protocol)
+
+                original_report_status = update_handler._report_status
+                update_handler._report_status = Mock(side_effect=original_report_status)
+
+                is_running_patch.start()
+                yield update_handler
+        finally:
+            is_running_patch.stop()
+
+    @staticmethod
+    def _goal_state_handler(url, **kwargs): # pylint: disable=unused-argument
+        """
+        Note: this function's signature has to coincide with that of http_get_request,
+        and thus needs **kwargs. We don't need the information, though, so we tell pylint 
+        we aren't planning on using the argument.
+
+        Relies on this function to have the property return_vals, so it'll need to be
+        set before being called
+        """
+        if not HttpRequestPredicates.is_goal_state_request(url):
+            return None
+        try:
+            return ReportStatusTestCase._goal_state_handler.return_vals.pop()
```

Comment:
From the new tests that are using this function, all of them are pretty much returning None. Why not keep it simple and remove the `return_vals` property and just return `None` in all cases?

Reply:
I think the main issue here is that we need to save state between function calls, so the static property would I think be needed regardless. I.e. replacing this line with `return None` would just require us to use a static property `count` or something that we decrement each time.

That's obviously do-able, but I'm not sure it really gains much. I could be missing something simple, though, did you have a specific idea on how to do this?

Reply:
I dont get it sorry. From the code usage, you're only returning `None` as the return_vals, I didnt see any other return value at all anywhere. Am I missing something here?

Reply:
Yes, I think this function is not clear enough and I'll need to either do some refactoring or add more comments. The "`other return value`" is actually `raise HttpError()`.

The idea behind this method is to return a list of goal states as requested, in order, and then fail with an `HttpError`, mimicking the scenario this PR is supposed to fix. In both of the tests below, I've written it to just return the same goal state either two or three times before failing with `HttpError`, but as written, this concievably supports a different goal state each time.